File: backend/app/schemas/prediction.py
"""Schemas Pydantic de request/response."""

from pydantic import BaseModel, Field


class BoundingBox(BaseModel):
    x1: float
    y1: float
    x2: float
    y2: float


class NormalizedBBox(BaseModel):
    x1: float = Field(ge=0.0, le=1.0)
    y1: float = Field(ge=0.0, le=1.0)
    x2: float = Field(ge=0.0, le=1.0)
    y2: float = Field(ge=0.0, le=1.0)


class Detection(BaseModel):
    class_id: int
    class_name: str
    confidence: float = Field(ge=0.0, le=1.0)
    bbox: BoundingBox
    bbox_normalized: NormalizedBBox | None = None
    # Origen del crack tras combinar full-frame + ROI: "full_frame" | "egg_roi" | "both".
    crack_source: str | None = None


class PrimaryEgg(BaseModel):
    confidence: float = Field(ge=0.0, le=1.0)
    bbox: BoundingBox
    bbox_normalized: NormalizedBBox


class ImageInfo(BaseModel):
    width: int = Field(ge=1)
    height: int = Field(ge=1)


class ModelInfo(BaseModel):
    name: str
    imgsz: int


class PredictionResponse(BaseModel):
    status: str
    route: str
    reason: str
    egg_detected: bool
    crack_detected: bool
    image: ImageInfo
    primary_egg: PrimaryEgg | None = None
    cracks: list[Detection] = Field(default_factory=list)
    raw_detections: list[Detection]
    detections: list[Detection]
    inference_ms: float = Field(ge=0.0)
    model: ModelInfo

    # Diagnóstico de la segunda etapa (ROI) — no consumido por la UI final.
    roi_used: bool = False
    roi_bbox: BoundingBox | None = None
    roi_cracks: list[Detection] = Field(default_factory=list)
    best_roi_crack_confidence: float | None = None
    crack_source: str | None = None
    full_frame_inference_ms: float | None = None
    roi_inference_ms: float | None = None
    total_inference_ms: float | None = None


class HealthResponse(BaseModel):
    status: str
    service: str
    model_loaded: bool
    model: str
    input_size: int
    model_path: str | None = None
    model_error: str | None = None


class RootResponse(BaseModel):
    name: str
    version: str
    docs: str

File: mobile/src/types/prediction.ts
/** Estados de clasificación devueltos por el backend (raw/debug). */
export type DetectionStatus = 'approved' | 'rejected' | 'unknown';

/** Ruta sugerida por el backend para el flujo operativo. */
export type DetectionRoute = 'accept' | 'reject' | 'review';

export type DetectionReason =
  | 'no_crack_detected'
  | 'crack_detected'
  | 'egg_not_detected'
  | 'crack_without_confirmed_egg';

/** Estado único para UX. */
export type DisplayStatus = 'detecting' | 'approved' | 'rejected' | 'review';

export interface BBox {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export interface NormalizedBBox {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export interface Detection {
  class_id: number;
  class_name: string;
  confidence: number;
  bbox: BBox;
  bbox_normalized?: NormalizedBBox;
}

export interface PrimaryEgg {
  confidence: number;
  bbox: BBox;
  bbox_normalized: NormalizedBBox;
}

export interface ImageInfo {
  width: number;
  height: number;
}

export interface ModelInfo {
  name: string;
  imgsz: number;
}

export interface PredictionResponse {
  status: DetectionStatus;
  route: DetectionRoute;
  reason: DetectionReason;
  egg_detected: boolean;
  crack_detected: boolean;
  image: ImageInfo;
  /** Fuente única de verdad para el huevo visible. */
  primary_egg?: PrimaryEgg | null;
  /** Cracks espacialmente asociados al primary_egg. */
  cracks?: Detection[];
  raw_detections?: Detection[];
  detections: Detection[];
  inference_ms: number;
  model: ModelInfo;
}

export interface HealthResponse {
  status: string;
  service: string;
  model_loaded: boolean;
  model: string;
  input_size: number;
  model_path?: string | null;
  model_error?: string | null;
}
